[JSC] Allocate b3 types using SequesteredArenaAllocator

Achierius opened the pull request:
#### 4f8543eb59c333c82931a72c54759848f0b7cf12
<pre>
[JSC] Allocate b3 types using SequesteredArenaAllocator
<a href="https://bugs.webkit.org/show_bug.cgi?id=287995">https://bugs.webkit.org/show_bug.cgi?id=287995</a>
<a href="https://rdar.apple.com/145159538">rdar://145159538</a>

Reviewed by Yijia Huang.

This covers all of the air-internal types whose lifetimes do not extend
past that of a compilation job. Allocating these out of the arena is
intended to improve security and performance.
Note that this does not yet cover all bytes used by Air: components
which themselves allocate (e.g. Vector) need to be handled separately.
This patch just ensures that when the top-level types are allocated
directly, they&apos;re allocated using SequesteredArenaMalloc.

The primary notable type not covered is the B3DataSection, which will
need to be allocated using the future persistent allocator.

* Source/JavaScriptCore/b3/B3BackwardsCFG.h:
* Source/JavaScriptCore/b3/B3BackwardsDominators.h:
* Source/JavaScriptCore/b3/B3BasicBlock.cpp:
* Source/JavaScriptCore/b3/B3BasicBlock.h:
* Source/JavaScriptCore/b3/B3CFG.h:
* Source/JavaScriptCore/b3/B3CheckSpecial.cpp:
* Source/JavaScriptCore/b3/B3CheckSpecial.h:
* Source/JavaScriptCore/b3/B3DataSection.h:
* Source/JavaScriptCore/b3/B3Dominators.h:
* Source/JavaScriptCore/b3/B3NaturalLoops.h:
* Source/JavaScriptCore/b3/B3PhiChildren.cpp:
* Source/JavaScriptCore/b3/B3PhiChildren.h:
* Source/JavaScriptCore/b3/B3Procedure.cpp:
* Source/JavaScriptCore/b3/B3Procedure.h:
* Source/JavaScriptCore/b3/B3StackmapSpecial.h:
* Source/JavaScriptCore/b3/B3TZoneImpls.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueRep.cpp:
* Source/JavaScriptCore/b3/B3ValueRep.h:
* Source/JavaScriptCore/b3/B3Variable.cpp:
* Source/JavaScriptCore/b3/B3Variable.h:
* Source/JavaScriptCore/ftl/FTLState.cpp:
(JSC::FTL::State::State):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::parseAndCompileOMG):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::parseAndCompileOMG):

Canonical link: <a href="https://commits.webkit.org/291921@main">https://commits.webkit.org/291921@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9296ac6b75a6bb84e8414f06943fe68702611f1f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94364 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13956 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3730 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99383 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44898 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14254 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22384 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72021 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29351 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97366 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10604 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/85219 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52351 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10295 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2900 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44212 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/87075 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80531 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3001 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101427 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/93031 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21419 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15635 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81025 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21670 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81227 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80397 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20042 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24947 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2321 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14646 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21402 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26583 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115680 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21090 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24551 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22831 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 reviewed (approved):
r=me if bots are happy.

Achierius committed:
[JSC] Allocate b3 types using SequesteredArenaAllocator
https://bugs.webkit.org/show_bug.cgi?id=287995
rdar://145159538

Reviewed by Yijia Huang.

This covers all of the air-internal types whose lifetimes do not extend
past that of a compilation job. Allocating these out of the arena is
intended to improve security and performance.
Note that this does not yet cover all bytes used by Air: components
which themselves allocate (e.g. Vector) need to be handled separately.
This patch just ensures that when the top-level types are allocated
directly, they're allocated using SequesteredArenaMalloc.

The primary notable type not covered is the B3DataSection, which will
need to be allocated using the future persistent allocator.

* Source/JavaScriptCore/b3/B3BackwardsCFG.h:
* Source/JavaScriptCore/b3/B3BackwardsDominators.h:
* Source/JavaScriptCore/b3/B3BasicBlock.cpp:
* Source/JavaScriptCore/b3/B3BasicBlock.h:
* Source/JavaScriptCore/b3/B3CFG.h:
* Source/JavaScriptCore/b3/B3CheckSpecial.cpp:
* Source/JavaScriptCore/b3/B3CheckSpecial.h:
* Source/JavaScriptCore/b3/B3DataSection.h:
* Source/JavaScriptCore/b3/B3Dominators.h:
* Source/JavaScriptCore/b3/B3NaturalLoops.h:
* Source/JavaScriptCore/b3/B3PhiChildren.cpp:
* Source/JavaScriptCore/b3/B3PhiChildren.h:
* Source/JavaScriptCore/b3/B3Procedure.cpp:
* Source/JavaScriptCore/b3/B3Procedure.h:
* Source/JavaScriptCore/b3/B3StackmapSpecial.h:
* Source/JavaScriptCore/b3/B3TZoneImpls.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueRep.cpp:
* Source/JavaScriptCore/b3/B3ValueRep.h:
* Source/JavaScriptCore/b3/B3Variable.cpp:
* Source/JavaScriptCore/b3/B3Variable.h:
* Source/JavaScriptCore/ftl/FTLState.cpp:
(JSC::FTL::State::State):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::parseAndCompileOMG):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::parseAndCompileOMG):

Canonical link: https://commits.webkit.org/291921@main
